### PR TITLE
Adds the Transform Flags concept for tree transformations

### PR DESCRIPTION
**Pull request by rbuckton**
To ensure tree transformations in the TypeScript compiler are efficient, we compute information about various required transformations for each node. Transform flags are aggregated in a bottom-up fashion, so that we already have information about a node's subtree when we compute the transform flags for the node. These transform flags are initially computed during the **bind** phase of the compiler, as at that point we guarantee a full walk of the tree.

transformflags

The values in the `TransformFlags` enum fall into several categories:

* *Facts* - State something about a node or its subtree. Facts that start with "Contains" indicate that a node somewhere within the current node's subtree requires a transformation. Facts that do not start with "Contains" indicate that the current node itself requires a transformation.
  * *Markers* - A special class of *Facts* that indicate specific transformations that need to occur higher in the source tree.
* *Assertions* - Sets of *Facts* to apply to a node. These are used primarily in the `computeTransformFlagsForNode` function in `binder.ts`.
* *Exclusions* - Bitmasks that are used to prevent some *Facts* from being copied up the subtree.

As a result, the `Node` interface is extended to include two additional properties:
* `transformFlags` - Contains the `TransformFlags` that pertain to this specific node.
* `excludeTransformFlags` - Contains a bitmask of `TransformFlags` that should be excluded from this node's `transformFlags` when aggregating the transform flags of a subtree containing this node.

The binder is modified to compute the transform flags for each node as it walks the tree. As a performance optimization, we skip transform flag aggregation for ambient nodes, type nodes, and declaration files. 

Each transformation phase will likely introduce new nodes. As a result, the `visitNode` and `visitNodes` functions in `visitor.ts` perform supplemental aggregation of transform flags to ensure successive transformation phases have the requisite information to perform additional transformations. 

For example, it is possible that during the transformation from TypeScript to ES6 we may introduce an arrow function a for more idiomatic ES6 emit. This may then result in a new `this` capture which we would then need to know about during the next transformation. Rather than send the new tree through another pass of the checker, we will instead leverage the `TransformFlags.ContainsLexicalThis` and `TransformFlags.ContainsCapturedLexicalThis` flags to determine whether we need further transformation.

These flags will enable us to have more efficient transformers in the long term, that can make decisions about whether to continue to walk a subtree based on querying these flags. The following example builds on the one found in #6892:

```ts
function visitor(node: Node) {
  if (node.transformFlags & TransformFlags.ES7) {
    // This node is an ES7 node and needs a specific transformation
    return visitorWorker(node);
  }
  else if (node.transformFlags & TransformFlags.ContainsES7) {
    // This node contains an ES7 node somewhere in its subtree, so we must walk the subtree
    return visitEachChild(node, visitor);
  }
  else {
    // Neither this node, nor its subtree, require any transformation, so we can simply 
    // reuse the node and its subtree and avoid a costly walk
    return node;
  }
}

function visitorWorker(node: Node) {
  switch (node.kind) {
    case SyntaxKind.BinaryExpression:
      return visitBinaryExpression(<BinaryExpression>node);
    ...
    default:
      return visitEachChild(node, visitor);
  }
}

function visitBinaryExpression(node: BinaryExpression) {
  if (node.operatorToken.kind === SyntaxKind.AsteriskAsteriskToken) {
    return createCall(
      createPropertyAccess(createIdentifier("Math"), "pow"),
      [
        visitNode(node.left, visitor, isExpressionNode),
        visitNode(node.right, visitor, isExpressionNode)
      ]
    );
  }
  else {
    return visitEachChild(node, visitor);
  }
}
```

**Related Pull Requests:**
* #6892 - Adds the visitor API foundation for tree transformations.
* #6983 - **Adds the Transform Flags concept for tree transformations.**
* #6986 - Adds the transformFiles API for tree transformations.
* #6987 - Adds a simplified pretty printer for tree transformations.
* #7191 - Adds the TypeScript transformer.
* #7192 - Adds the JSX transformer.
* #7193 - Adds the ES7 transformer.
* #7194 - Adds the ES6 transformer.
* #7206 - Adds the Module transformer.

### CODE REVIEW

**Comment by rbuckton** [comment]
Paging for review: @mhegazy, @ahejlsberg, @yuit, @DanielRosenwasser, @RyanCavanaugh, @vladima

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -1751,4 +1791,437 @@ namespace ts {
...
+        let transformFlags: TransformFlags;
+        let excludeFlags: TransformFlags;
+        switch (node.kind) {
+            case SyntaxKind.PublicKeyword:
+            case SyntaxKind.PrivateKeyword:
+            case SyntaxKind.ProtectedKeyword:
+            case SyntaxKind.AbstractKeyword:
+            case SyntaxKind.DeclareKeyword:
+            case SyntaxKind.AsyncKeyword:
+            case SyntaxKind.ConstKeyword:
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: Does this trigger false positives for `const x = ...` variable declarations?

Comment (rbuckton) [inline src/compiler/binder.ts]: No, we don't treat `const` as a modifier for variable declarations.

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -1751,4 +1791,437 @@ namespace ts {
...
+                if (!nodeIsSynthesized(node)) {
+                    if ((<ParenthesizedExpression>node).expression.kind === SyntaxKind.AsExpression
+                        || (<ParenthesizedExpression>node).expression.kind === SyntaxKind.TypeAssertionExpression) {
+                        transformFlags = TransformFlags.AssertTypeScript;
+                    }
+                }
+                break;
+
+            case SyntaxKind.ExpressionStatement:
+                // if (node.flags & NodeFlags.Generated) {
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: Remove / explain why it's commented out

File: src/compiler/binder.ts [hunk 3]
```diff
@@ -1751,4 +1791,437 @@ namespace ts {
...
+                // If a VariableDeclarationList is `let` or `const`, then it is ES6 syntax.
+                if (node.flags & NodeFlags.Let
+                    || node.flags & NodeFlags.Const) {
+                    transformFlags |= TransformFlags.AssertES6;
+                }
+
+                break;
+
+            case SyntaxKind.VariableStatement:
+                // If a VariableStatement is exported, then it is ES6 syntax.
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: This should only be true at top-level (i.e. not inside `namespace` declarations)

File: src/compiler/binder.ts [hunk 4]
```diff
@@ -1751,4 +1791,437 @@ namespace ts {
...
+                if (subtreeFlags & TransformFlags.ContainsParameterPropertyAssignments
+                    || subtreeFlags & TransformFlags.ContainsPropertyInitializer
+                    || subtreeFlags & TransformFlags.ContainsDecorators) {
+                    transformFlags |= TransformFlags.AssertTypeScript;
+                }
+
+                break;
+
+            case SyntaxKind.HeritageClause:
+                // An `extends` HertiageClause is ES6 syntax.
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: spelling (Hertiage -> Heritage)

File: src/compiler/binder.ts [hunk 5]
```diff
@@ -1751,4 +1791,437 @@ namespace ts {
...
+
+                break;
+
+            case SyntaxKind.MethodDeclaration:
+                // A MethodDeclaration is ES6 syntax.
+                excludeFlags = TransformFlags.MethodOrAccessorExcludes;
+                transformFlags |= TransformFlags.AssertES6;
+
+                // A MethodDeclaration is TypeScript syntax if it is either async, abstract, overloaded,
+                // generic, or has both a computed property name and a decorator.
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: Do `private` / `protected` fall in this category as well?

File: src/compiler/binder.ts (hunk 2 above)
Comment (rbuckton) [inline src/compiler/binder.ts]: I will have a later PR that uses this logic, but just didn't want it to get lost in the shuffle.

File: src/compiler/binder.ts (hunk 3 above)
Comment (rbuckton) [inline src/compiler/binder.ts]: Yes, but there is no way to know that when building these flags bottom-up. Instead, we at least assume it is ES6 syntax. When transforming a TypeScript namespace, we can easily look at each declaration that is an immediate child and handle those cases directly.

File: src/compiler/binder.ts [hunk 6]
```diff
@@ -1751,4 +1791,437 @@ namespace ts {
...
+                break;
+
+            case SyntaxKind.HeritageClause:
+                // An `extends` HertiageClause is ES6 syntax.
+                if ((<HeritageClause>node).token === SyntaxKind.ExtendsKeyword) {
+                    transformFlags |= TransformFlags.AssertES6;
+                }
+
+                // An `implements` HeritageClause is TypeScript syntax.
+                else if ((<HeritageClause>node).token === SyntaxKind.ImplementsKeyword) {
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts]: This should potentially be a `Debug.assert`

File: src/compiler/types.ts [hunk 7]
```diff
@@ -2734,6 +2736,52 @@ namespace ts {
...
+        ContainsDecorators = 1 << 8,
+        ContainsPropertyInitializer = 1 << 9,
+        ContainsLexicalThis = 1 << 10,
+        ContainsCapturedLexicalThis = 1 << 11,
+        ContainsDefaultValueAssignments = 1 << 12,
+        ContainsParameterPropertyAssignments = 1 << 13,
+        ContainsSpreadElementExpression = 1 << 14,
+        ContainsComputedPropertyName = 1 << 15,
+
+        // Assertions
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: Do these assertions actually get acted on in a later PR?

File: src/compiler/types.ts [hunk 8]
```diff
@@ -2734,6 +2736,52 @@ namespace ts {
         resolveModuleNames?(moduleNames: string[], containingFile: string): ResolvedModule[];
     }
 
+    /* @internal */
+    export const enum TransformFlags {
+        // Facts
+        // - Flags used to indicate that a node or subtree contains syntax that requires transformation.
+        TypeScript = 1 << 0,
+        ContainsTypeScript = 1 << 1,
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: What is a context in which `ContainsTypeScript` doesn't suffice for `TypeScript` itself.

File: src/compiler/types.ts (hunk 7 above)
Comment (rbuckton) [inline src/compiler/types.ts]: Yes, and the example above shows how they are used. The bitmask of an assertion is merely a shortcut used to ensure that both the "this node is a" flag (e.g. `TypeScript`) and "the subtree containing this node has a" flag (e.g. `ContainsTypeScript`) are set on the node.

File: src/compiler/types.ts (hunk 8 above)
Comment (rbuckton) [inline src/compiler/types.ts]: See the example above, the two flags are used separately to determine the correct course of action to take.

File: src/compiler/binder.ts [hunk 9]
```diff
@@ -1751,4 +1791,440 @@ namespace ts {
...
+     * Computes the transform flags for a node, given the transform flags of its subtree
+     *
+     * @param node The node to analyze
+     * @param subtreeFlags Transform flags computed for this node's subtree
+     */
+    export function computeTransformFlagsForNode(node: Node, subtreeFlags: TransformFlags) {
+        // Ambient nodes are TypeScript syntax and the flags of their subtree are ignored.
+        if (node.flags & NodeFlags.Ambient) {
+            return (node.transformFlags = TransformFlags.AssertTypeScript)
+                & ~(node.excludeTransformFlags = TransformFlags.NodeExcludes);
```
Comment (yuit) [inline src/compiler/binder.ts]: Not sure I understand why there is 'tilde' here?

Comment (weswigham) [inline src/compiler/binder.ts]: Bitwise not?

File: src/compiler/binder.ts (hunk 1 above)
Comment (yuit) [inline src/compiler/binder.ts]: how is 'const' has flag of `AssertTypeScript` should it be under `es6`?

File: src/compiler/binder.ts [hunk 10]
```diff
@@ -1751,4 +1791,440 @@ namespace ts {
...
+                break;
+
+            case SyntaxKind.NoSubstitutionTemplateLiteral:
+            case SyntaxKind.TemplateHead:
+            case SyntaxKind.TemplateMiddle:
+            case SyntaxKind.TemplateTail:
+            case SyntaxKind.TemplateExpression:
+            case SyntaxKind.TaggedTemplateExpression:
+            case SyntaxKind.ShorthandPropertyAssignment:
+            case SyntaxKind.ForOfStatement:
```
Comment (yuit) [inline src/compiler/binder.ts]: Should there be `ForInStatement`?

File: src/compiler/binder.ts [hunk 11]
```diff
@@ -1751,4 +1791,440 @@ namespace ts {
...
+                transformFlags |= TransformFlags.AssertES6;
+
+                // A MethodDeclaration is TypeScript syntax if it is either async, abstract, overloaded,
+                // generic, or has both a computed property name and a decorator.
+                if ((<MethodDeclaration>node).body === undefined
+                    || (<MethodDeclaration>node).typeParameters !== undefined
+                    || node.flags & NodeFlags.Async
+                    || node.flags & NodeFlags.Abstract
+                    || (subtreeFlags & TransformFlags.ContainsDecorators
+                        && subtreeFlags & TransformFlags.ContainsComputedPropertyName)) {
```
Comment (yuit) [inline src/compiler/binder.ts]: why you care about `ContainsComputedPropertyName` to set the flag with `AssertTypeScript`?

File: src/compiler/binder.ts [hunk 12]
```diff
@@ -1751,4 +1791,440 @@ namespace ts {
...
+            case SyntaxKind.GetAccessor:
+            case SyntaxKind.SetAccessor:
+                // A GetAccessor or SetAccessor is ES5 syntax.
+                excludeFlags = TransformFlags.MethodOrAccessorExcludes;
+
+                // A GetAccessor or SetAccessor is TypeScript syntax if it is either abstract,
+                // or has both a computed property name and a decorator.
+                if (node.flags & NodeFlags.Abstract ||
+                    subtreeFlags & TransformFlags.ContainsDecorators &&
+                    subtreeFlags & TransformFlags.ContainsComputedPropertyName) {
```
Comment (yuit) [inline src/compiler/binder.ts]: why you care about ContainsComputedPropertyName to set the flag with AssertTypeScript?

File: src/compiler/binder.ts [hunk 13]
```diff
@@ -1751,4 +1791,440 @@ namespace ts {
...
+                // If the PropertyDeclaration has an initializer, we need to inform its ancestor
+                // so that it handle the transformation.
+                if ((<PropertyDeclaration>node).initializer) {
+                    transformFlags |= TransformFlags.ContainsPropertyInitializer;
+                }
+
+                break;
+
+            case SyntaxKind.MethodDeclaration:
+                // A MethodDeclaration is ES6 syntax.
```
Comment (yuit) [inline src/compiler/binder.ts]: you don't need to capture whether the method is static or instance? Wouldn't that affect the emitting?

File: src/compiler/binder.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/binder.ts]: The only ES6 `const` is for `VariableDeclarationList`, which is not stored as a modifier. The only time `const` is a `Modifier` is for a const enum, which is TypeScript only.

File: src/compiler/binder.ts (hunk 9 above)
Comment (rbuckton) [inline src/compiler/binder.ts]: Yes, `~` gets the ones-complement of the operand. The result here is the `transformFlags` of the node, excluding `excludeTransformFlags` of the node.

File: src/compiler/binder.ts (hunk 10 above)
Comment (rbuckton) [inline src/compiler/binder.ts]: `ForInStatement` is ES3, so there's currently no specific transformation associated with it.

File: src/compiler/binder.ts (hunk 11 above)
Comment (rbuckton) [inline src/compiler/binder.ts]: This has to do with decorators. We don't want to re-evaluate the expression of the computed property name:

```ts
// ts
class C {
  @dec
  [generateRandomMethodName()]() {
  }
}

// js
class C {
  [_a = generateRandomMethodName()]() {
  }
}
__decorate([dec], C.prototype, _a, null);
```

We need to flag the method itself as needing a specific transformation in this case, to transform the computed property name.

File: src/compiler/binder.ts (hunk 12 above)
Comment (rbuckton) [inline src/compiler/binder.ts]: See https://github.com/Microsoft/TypeScript/pull/6983#discussion_r53099211 above for rationale.

File: src/compiler/binder.ts (hunk 13 above)
Comment (rbuckton) [inline src/compiler/binder.ts]: We can test the node's `flags` for this, the transformation flags just give us a very high level indication that something needs to happen.

File: src/compiler/binder.ts (hunk 9 above)
Comment (yuit) [inline src/compiler/binder.ts]: Gotcha . Thanks @weswigham and @rbuckton. Also could you put return type in the signature?

File: src/compiler/binder.ts (hunk 10 above)
Comment (yuit) [inline src/compiler/binder.ts]: Oh right. Sorry, got it mixed up

**Pull request merged**